lighthouse-core/gather/gatherers/dobetterweb/optimized-images.js

```diff
@@ -0,0 +1,128 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+ /**
+  * @fileoverview Determines optimized jpeg/webp filesizes for all same-origin and dataURI images.
+  */
+'use strict';
+
+const Gatherer = require('../gatherer');
+const URL = require('../../../lib/url-shim');
+
+/* global document */
+
+/* istanbul ignore next */
+function getOptimizedNumBytes(url) {
+  return new Promise(function(resolve, reject) {
+    const img = new Image();
+    const canvas = document.createElement('canvas');
+    const context = canvas.getContext('2d');
+
+    function getTypeStats(type, quality) {
+      const dataURI = canvas.toDataURL(type, quality);
+      const base64 = dataURI.slice(dataURI.indexOf(',') + 1);
+      return {base64: base64.length, binary: atob(base64).length};
+    }
+
+    document.body.appendChild(canvas);
+    img.addEventListener('load', () => {
+      try {
+        canvas.height = img.height;
+        canvas.width = img.width;
+        context.drawImage(img, 0, 0);
+
+        const jpeg = getTypeStats('image/jpeg', 0.8);
+        const webp = getTypeStats('image/webp', 0.8);
+
+        document.body.removeChild(canvas);
+        resolve({jpeg, webp});
+      } catch (err) {
+        reject(err);
+      }
+    }, false);
+
+    img.src = url;
+  });
+}
+
+class OptimizedImages extends Gatherer {
+  /**
+   * @param {string} pageUrl
+   * @param {NetworkRecords} networkRecords
+   * @return {!Array.<!{url: string, isBase64DataUri: boolean, mimeType: string, resourceSize: number}>}
+   */
+  static filterImageRequests(pageUrl, networkRecords) {
+    return networkRecords.reduce((prev, record) => {
+      const isOptimizableImage = /image\/(png|bmp|jpeg)/.test(record._mimeType);
+      const isSameOrigin = URL.hostsMatch(pageUrl, record._url);
+      const isBase64DataUri = /^data.{2,40}base64/.test(record._url);
```

Comment:
done, are there cases where a URL could begin with `^data:` and not be a data URI? don't want to make it brittle/less readable with a bunch of whitespace matchers if it doesn't matter